File: python/telequick_agents/agent.py
"""High-level external-agent API: register-and-wait, handle calls.

The drop-in seam for every provider starter in this repo. Your process dials
ONE outbound QUIC/WebTransport session to the engine, authenticates with your
media app key/secret, registers under ``agent_name``, and gets a ``Call`` for
every phone call the platform routes to it:

    import asyncio
    from telequick_agents import AgentConfig, serve

    async def handler(call):
        print("call from", call.caller_number)
        async for pcm in call.audio():          # 20 ms pcm16 @ 8 kHz
            await call.send_audio(pcm)          # echo

    asyncio.run(serve(handler, AgentConfig.from_env(agent_name="salesbot")))

Route a number to the agent by pointing a platform agent's ``VENDOR_BRIDGE``
node (``vendor_room = agent_name``) at it — the console's "External agent"
create flow does this and mints the media key in one step.
"""

from __future__ import annotations

import asyncio
import logging
import os
from dataclasses import dataclass
import inspect
from typing import AsyncIterator, Awaitable, Callable, Optional

from .transport import PlayoutReport, WebTransportConfig, WebTransportSession
from .types import JobAssign

logger = logging.getLogger("telequick")


@dataclass
class AgentConfig:
    """Everything the worker needs to reach the engine and identify itself."""

    host: str
    app_key: str
    app_secret: str
    agent_name: str
    port: int = 443
    sample_rate: int = 8000  # PSTN legs are 8 kHz
    verify: bool = True  # dev only: False skips engine cert validation

    @classmethod
    def from_env(cls, **overrides) -> "AgentConfig":
        """Build from TELEQUICK_HOST / TELEQUICK_MEDIA_KEY /
        TELEQUICK_MEDIA_SECRET / TELEQUICK_AGENT (overridable per kwarg)."""
        vals = dict(
            host=os.environ.get("TELEQUICK_HOST", ""),
            app_key=os.environ.get("TELEQUICK_MEDIA_KEY", ""),
            app_secret=os.environ.get("TELEQUICK_MEDIA_SECRET", ""),
            agent_name=os.environ.get("TELEQUICK_AGENT", ""),
        )
        vals.update(overrides)
        missing = [k for k in ("host", "app_key", "app_secret", "agent_name") if not vals.get(k)]
        if missing:
            raise ValueError(
                f"AgentConfig missing {missing} — set TELEQUICK_HOST/TELEQUICK_MEDIA_KEY/"
                "TELEQUICK_MEDIA_SECRET/TELEQUICK_AGENT or pass them explicitly"
            )
        return cls(**vals)

    def _wt(self, call_id: int = 0) -> WebTransportConfig:
        return WebTransportConfig(
            engine_url=f"https://{self.host}:{self.port}",
            app_key=self.app_key,
            app_secret=self.app_secret,
            agent_name=self.agent_name,
            call_id=call_id,
            sample_rate=self.sample_rate,
        )


class Call:
    """One live phone call handed to your worker.

    Audio in both directions is 16-bit PCM at ``sample_rate`` (8 kHz for
    telephony). Inbound arrives as 20 ms frames; outbound accepts any chunk
    size and is paced onto the wire as steady 20 ms frames.
    """

    def __init__(self, assign: JobAssign, session: WebTransportSession) -> None:
        self._assign = assign
        self._session = session
        self._ended = asyncio.Event()

    # -- identity ---------------------------------------------------------
    @property
    def call_id(self) -> int:
        return self._assign.call_id

    @property
    def room(self) -> str:
        return self._assign.room_name

    @property
    def caller_number(self) -> str:
        """ANI — who is calling."""
        return self._assign.caller_number

    @property
    def called_number(self) -> str:
        """DNIS — the number they dialed."""
        return self._assign.called_number

    @property
    def trunk_id(self) -> str:
        return self._assign.trunk_id

    @property
    def attributes(self) -> dict:
        return self._assign.attributes

    @property
    def tags(self) -> dict:
        """The agent's effective resource tags ({key: value}) — the platform's
        governed cost-allocation / ownership / environment labels, inherited
        down the parent chain. Use them to route, bill, or branch per tenant
        without a second lookup."""
        return self._assign.tags

    @property
    def sample_rate(self) -> int:
        return self._session.sample_rate

    @property
    def ended(self) -> bool:
        return self._ended.is_set()

    # -- audio ------------------------------------------------------------
    async def recv_audio(self) -> bytes | None:
        """Next caller frame (pcm16), or None once the call ends."""
        pcm = await self._session.recv_frame()
        if pcm is None:
            self._ended.set()
        return pcm

    async def audio(self) -> AsyncIterator[bytes]:
        """Iterate caller audio until hangup: ``async for pcm in call.audio()``."""
        while True:
            pcm = await self.recv_audio()
            if pcm is None:
                return
            yield pcm

    async def send_audio(self, pcm: bytes) -> None:
        """Queue agent audio for the caller (pcm16 @ ``sample_rate``)."""
        await self._session.send_frame(pcm)

    async def flush(self) -> None:
        """Mark the current outbound segment complete (enables playout reports)."""
        await self._session.signal_flush()

    async def clear(self) -> None:
        """Barge-in: drop buffered agent audio so playout stops immediately."""
        await self._session.signal_clear()

    async def next_playout(self) -> PlayoutReport:
        return await self._session.next_playout()

    # -- platform parity --------------------------------------------------
    def send_transcript(self, role: str, text: str, is_final: bool = True) -> None:
        """Forward a turn ('user'/'assistant') into platform transcript
        storage, analytics, and ``voice.transcript.ready`` webhooks."""
        self._session.send_transcript(role, text, is_final)

    async def aclose(self) -> None:
        self._ended.set()
        await self._session.aclose()


Handler = Callable[[Call], Awaitable[None]]
# Lifecycle hooks — sync or async both work.
ReadyHook = Callable[[], object]
DisconnectHook = Callable[[Exception], object]


async def _fire(cb: Optional[Callable], *args) -> None:
    if cb is None:
        return
    try:
        res = cb(*args)
        if inspect.isawaitable(res):
            await res
    except Exception:  # noqa: BLE001 — a bad hook must never take the worker down
        logger.exception("telequick: lifecycle hook raised")


async def serve(
    handler: Handler,
    cfg: AgentConfig,
    *,
    on_ready: Optional[ReadyHook] = None,
    on_disconnected: Optional[DisconnectHook] = None,
) -> None:
    """Connect, authenticate, register, and run ``handler`` per assigned call.

    Reconnects with exponential backoff on any drop (engine restart, network),
    so presence recovers without operator action. Runs until cancelled.

    To know the worker is live, pass ``on_ready`` (fires when connected +
    registered, and again after every reconnect) and ``on_disconnected`` (fires
    on a drop, before the retry). Both may be sync or async.
    """
    from .quic_client import media_connection

    async def _serve_one(assign: JobAssign, session: WebTransportSession) -> None:
        call = Call(assign, session)
        try:
            await handler(call)
        except Exception:  # noqa: BLE001
            logger.exception("telequick: handler failed for call_id=%s", assign.call_id)
        finally:
            if not session.closed:
                await call.aclose()

    backoff = 1.0
    while True:
        try:
            async with media_connection(
                cfg._wt(), host=cfg.host, port=cfg.port, verify=cfg.verify
            ) as mc:
                logger.info("telequick worker ready (agent=%s) — awaiting calls", cfg.agent_name)
                backoff = 1.0
                await _fire(on_ready)
                closed = getattr(mc, "closed_event", None)
                tasks: set[asyncio.Task] = set()
                try:
                    while True:
                        next_task = asyncio.ensure_future(mc.next_call())
                        waiters = [next_task]
                        closed_task = asyncio.ensure_future(closed.wait()) if closed else None
                        if closed_task is not None:
                            waiters.append(closed_task)
                        await asyncio.wait(waiters, return_when=asyncio.FIRST_COMPLETED)
                        if closed is not None and closed.is_set():
                            next_task.cancel()
                            if closed_task is not None:
                                closed_task.cancel()
                            raise ConnectionError("engine connection closed")
                        if closed_task is not None:
                            closed_task.cancel()
                        assign, session = next_task.result()
                        logger.info(
                            "telequick: call_id=%s from=%s to=%s → handler",
                            assign.call_id, assign.caller_number, assign.called_number,
                        )
                        t = asyncio.create_task(_serve_one(assign, session))
                        tasks.add(t)
                        t.add_done_callback(tasks.discard)
                finally:
                    for t in tasks:
                        t.cancel()
        except asyncio.CancelledError:
            raise
        except Exception as e:  # noqa: BLE001 — reconnect on any transport failure
            await _fire(on_disconnected, e)
            logger.warning(
                "telequick worker connection lost (%s: %s) — reconnecting in %.1fs",
                type(e).__name__, e, backoff,
            )
            await asyncio.sleep(backoff)
            backoff = min(backoff * 2, 30.0)


def run(
    handler: Handler,
    cfg: AgentConfig | None = None,
    *,
    on_ready: Optional[ReadyHook] = None,
    on_disconnected: Optional[DisconnectHook] = None,
) -> None:
    """Blocking convenience wrapper: ``run(handler)`` with env-based config."""
    asyncio.run(
        serve(handler, cfg or AgentConfig.from_env(), on_ready=on_ready, on_disconnected=on_disconnected)
    )

File: agent.go
package agents

import (
	"context"
	"errors"
	"fmt"
	"iter"
	"log"
	"os"
	"strconv"
	"time"
)

// AgentConfig is everything the worker needs to reach the engine and identify
// itself.
type AgentConfig struct {
	Host       string
	AppKey     string
	AppSecret  string
	AgentName  string
	Port       int  // default 443
	SampleRate int  // default 8000 — PSTN legs are 8 kHz
	Verify     bool // default true; false skips engine cert validation (dev only)
}

// ConfigFromEnv builds an AgentConfig from TELEQUICK_HOST / TELEQUICK_MEDIA_KEY /
// TELEQUICK_MEDIA_SECRET / TELEQUICK_AGENT (and optional TELEQUICK_PORT).
// Any field set on `override` wins.
func ConfigFromEnv(override AgentConfig) (AgentConfig, error) {
	c := AgentConfig{
		Host:       os.Getenv("TELEQUICK_HOST"),
		AppKey:     os.Getenv("TELEQUICK_MEDIA_KEY"),
		AppSecret:  os.Getenv("TELEQUICK_MEDIA_SECRET"),
		AgentName:  os.Getenv("TELEQUICK_AGENT"),
		Port:       443,
		SampleRate: 8000,
		Verify:     true,
	}
	if p, err := strconv.Atoi(os.Getenv("TELEQUICK_PORT")); err == nil && p > 0 {
		c.Port = p
	}
	if override.Host != "" {
		c.Host = override.Host
	}
	if override.AppKey != "" {
		c.AppKey = override.AppKey
	}
	if override.AppSecret != "" {
		c.AppSecret = override.AppSecret
	}
	if override.AgentName != "" {
		c.AgentName = override.AgentName
	}
	if override.Port != 0 {
		c.Port = override.Port
	}
	if override.SampleRate != 0 {
		c.SampleRate = override.SampleRate
	}
	var missing []string
	for _, kv := range [][2]string{{"host", c.Host}, {"app key", c.AppKey}, {"app secret", c.AppSecret}, {"agent name", c.AgentName}} {
		if kv[1] == "" {
			missing = append(missing, kv[0])
		}
	}
	if len(missing) > 0 {
		return c, fmt.Errorf("AgentConfig missing %v — set TELEQUICK_HOST / TELEQUICK_MEDIA_KEY / TELEQUICK_MEDIA_SECRET / TELEQUICK_AGENT or pass them explicitly", missing)
	}
	return c, nil
}

func (c AgentConfig) wt() WebTransportConfig {
	return WebTransportConfig{
		EngineURL:   fmt.Sprintf("https://%s:%d", c.Host, c.Port),
		AppKey:      c.AppKey,
		AppSecret:   c.AppSecret,
		AgentName:   c.AgentName,
		SampleRate:  c.SampleRate,
		NumChannels: 1,
	}
}

// Call is one live phone call handed to your handler. Audio in both
// directions is 16-bit PCM at SampleRate (8 kHz for telephony). Inbound
// arrives as 20 ms frames; outbound accepts any chunk size and is paced onto
// the wire as steady 20 ms frames.
type Call struct {
	assign JobAssign
	sess   *WebTransportSession
}

func (c *Call) CallID() uint16                   { return c.assign.CallID }
func (c *Call) Room() string                     { return c.assign.RoomName }
func (c *Call) CallerNumber() string             { return c.assign.CallerNumber } // ANI
func (c *Call) CalledNumber() string             { return c.assign.CalledNumber } // DNIS
func (c *Call) TrunkID() string                  { return c.assign.TrunkID }
func (c *Call) Attributes() map[string]any       { return c.assign.Attributes }
func (c *Call) Metadata() map[string]string      { return c.assign.Metadata }
func (c *Call) SampleRate() int                  { return c.sess.SampleRate() }
func (c *Call) Ended() bool                      { return c.sess.Closed() }

// Tags are the agent's effective resource tags ({key: value}) — the
// platform's governed cost-allocation / ownership / environment labels,
// inherited down the parent chain. Route, bill, or branch per tenant without
// a second lookup.
func (c *Call) Tags() map[string]string { return c.assign.Tags }

// RecvAudio returns the next caller frame (pcm16); ok=false once the call ends.
func (c *Call) RecvAudio() ([]byte, bool) { return c.sess.RecvFrame() }

// Audio iterates caller audio until hangup: `for pcm := range call.Audio() { … }`.
func (c *Call) Audio() iter.Seq[[]byte] {
	return func(yield func([]byte) bool) {
		for {
			pcm, ok := c.sess.RecvFrame()
			if !ok || !yield(pcm) {
				return
			}
		}
	}
}

// SendAudio queues agent audio for the caller (pcm16 @ SampleRate).
func (c *Call) SendAudio(pcm []byte) { c.sess.SendFrame(pcm) }

// Flush marks the current outbound segment complete (enables playout reports).
func (c *Call) Flush() { c.sess.SignalFlush() }

// Clear is barge-in: drop buffered agent audio so playout stops immediately.
func (c *Call) Clear() { c.sess.SignalClear() }

// NextPlayout blocks for the next playout report.
func (c *Call) NextPlayout() PlayoutReport { return c.sess.NextPlayout() }

// SendTranscript forwards a turn ("user"/"assistant") into platform transcript
// storage, analytics, and voice.transcript.ready webhooks.
func (c *Call) SendTranscript(role, text string, isFinal bool) {
	c.sess.SendTranscript(role, text, isFinal)
}

// Close ends the call from the agent side.
func (c *Call) Close() { c.sess.Close() }

// Handler serves one call; return when the call is done.
type Handler func(ctx context.Context, call *Call) error

// ServeOptions carries the connection-lifecycle hooks.
type ServeOptions struct {
	// OnReady fires once the worker is connected, authenticated and registered
	// — present to the platform and awaiting calls — and again after every
	// automatic reconnect. Drive a readiness probe / health check off it.
	OnReady func()
	// OnDisconnected fires when the connection drops (engine restart,
	// network); Serve then reconnects with backoff and OnReady fires again.
	OnDisconnected func(error)
	// Logger defaults to the standard logger.
	Logger *log.Logger
}

// Serve connects, authenticates, registers, and runs handler once per assigned
// call. It reconnects with exponential backoff on any drop so presence recovers
// without operator action, and returns only when ctx is cancelled.
func Serve(ctx context.Context, handler Handler, cfg AgentConfig, opts ServeOptions) error {
	lg := opts.Logger
	if lg == nil {
		lg = log.Default()
	}
	backoff := time.Second
	for {
		err := serveOnce(ctx, handler, cfg, opts, lg)
		if ctx.Err() != nil {
			return ctx.Err()
		}
		if err == nil {
			err = errClosed
		}
		if opts.OnDisconnected != nil {
			opts.OnDisconnected(err)
		}
		lg.Printf("[telequick] worker connection lost (%v) — reconnecting in %s", err, backoff)
		select {
		case <-time.After(backoff):
		case <-ctx.Done():
			return ctx.Err()
		}
		if backoff *= 2; backoff > 30*time.Second {
			backoff = 30 * time.Second
		}
	}
}

func serveOnce(ctx context.Context, handler Handler, cfg AgentConfig, opts ServeOptions, lg *log.Logger) error {
	conn, err := OpenMediaConnection(ctx, cfg.wt(), cfg.Host, cfg.Port, cfg.Verify)
	if err != nil {
		return err
	}
	defer conn.Close()
	lg.Printf("[telequick] worker ready (agent=%s) — awaiting calls", cfg.AgentName)
	if opts.OnReady != nil {
		opts.OnReady()
	}

	cctx, cancel := context.WithCancel(ctx)
	defer cancel()
	go func() {
		select {
		case <-conn.Closed:
			cancel()
		case <-cctx.Done():
		}
	}()

	for {
		assign, sess, err := conn.MC.NextCall(cctx)
		if err != nil {
			select {
			case <-conn.Closed:
				return errClosed
			default:
				return err
			}
		}
		lg.Printf("[telequick] call_id=%d from=%s to=%s → handler", assign.CallID, assign.CallerNumber, assign.CalledNumber)
		go func(assign JobAssign, sess *WebTransportSession) {
			call := &Call{assign: assign, sess: sess}
			defer func() {
				if r := recover(); r != nil {
					lg.Printf("[telequick] handler panicked for call_id=%d: %v", assign.CallID, r)
				}
				if !sess.Closed() {
					call.Close()
				}
			}()
			if err := handler(cctx, call); err != nil && !errors.Is(err, context.Canceled) {
				lg.Printf("[telequick] handler failed for call_id=%d: %v", assign.CallID, err)
			}
		}(assign, sess)
	}
}
